core/trino-main/src/main/java/io/trino/operator/output/UnnestingPositionsAppender.java

```diff
@@ -94,4 +117,55 @@ private IntArrayList mapPositions(IntArrayList positions, DictionaryBlock block)
         }
         return IntArrayList.wrap(positionArray);
     }
+
+    private static class DictionaryBlockBuilder
+    {
+        private static final int INSTANCE_SIZE = toIntExact(ClassLayout.parseClass(DictionaryBlockBuilder.class).instanceSize());
+        private Block dictionary;
+        private int[] dictionaryIds = new int[0];
+        private int positionCount;
+        private boolean closed;
+
+        public boolean isEmpty()
+        {
+            return positionCount == 0;
+        }
+
+        public Block build()
+        {
+            return DictionaryBlock.create(positionCount, dictionary, dictionaryIds);
+        }
+
+        public long getRetainedSizeInBytes()
+        {
+            return INSTANCE_SIZE + (long) dictionaryIds.length * Integer.BYTES + (dictionary != null ? dictionary.getRetainedSizeInBytes() : 0);
+        }
+
+        public boolean canAppend(Block dictionary)
+        {
+            return !closed && (dictionary == this.dictionary || this.dictionary == null);
+        }
+
+        public void append(IntArrayList mappedPositions, Block dictionary)
+        {
+            checkArgument(canAppend(dictionary));
+            this.dictionary = dictionary;
+            dictionaryIds = IntArrays.ensureCapacity(dictionaryIds, positionCount + mappedPositions.size(), positionCount);
```

Comment:
Why not use `IntArrayList`?

Reply:
I need to calculate `getRetainedSizeInBytes`. For that, the actual size of the array is needed and `IntArrayList` does not expose that info.